osquery/events/windows/ntfs_event_publisher.cpp

```diff
@@ -0,0 +1,595 @@
+      std::string dir;
+      auto status = getPathFromReferenceNumber(
+          dir, journal_record.drive_letter, journal_record.node_ref_number);
+
+      if (!status.ok()) {
+        TLOG << "Failed to get directory for parent FRN cache update";
+      } else {
+        parent_frn_cache[journal_record.node_ref_number] = dir;
+      }
+      // Intentional fallthrough.
+    }
+    case USNJournalEventRecord::Type::FileRename_NewName: {
+      auto it = rename_path_mapper.find(journal_record.node_ref_number);
+      if (it == rename_path_mapper.end()) {
+        skip_record = true;
+        VLOG(1) << "Failed to remap the rename records";
+        break;
+      }
+
+      old_name_record = it->second;
+      rename_path_mapper.erase(it);
+    }
+    }
+
+    if (skip_record) {
+      continue;
+    }
+
+    // Generate the new event
+    NTFSEventRecord event(journal_record);
+
+    // TODO(ww): This is failing occasionally for files that do exist
+    // on disk, but only on the first call to look them up. I'm not
+    // sure why yet, but falling back on the parent FRN cache and
+    // building the path from it works for now.
+    // See https://github.com/osquery/osquery/issues/5848
+    auto status = getPathFromReferenceNumber(event.path,
+                                             journal_record.drive_letter,
+                                             journal_record.node_ref_number);
+    if (!status.ok()) {
+      TLOG << "FRN pathname lookup failed, trying parent: "
+           << status.getMessage();
+
+      status = getPathFromParentFRN(event.path,
+                                    parent_frn_cache,
+                                    journal_record.drive_letter,
+                                    journal_record.name,
+                                    journal_record.parent_ref_number);
+
+      if (!status.ok()) {
+        VLOG(1) << "Parent FRN lookup failed: " << status.getMessage();
+
+        event.path = journal_record.name;
+        event.partial = true;
+      }
+    }
+
+    if (old_name_record.node_ref_number != 0U) {
+      status = getPathFromParentFRN(event.old_path,
+                                    parent_frn_cache,
+                                    old_name_record.drive_letter,
+                                    old_name_record.name,
+                                    old_name_record.parent_ref_number);
+      if (!status.ok()) {
+        VLOG(1) << "Parent FRN lookup failed: " << status.getMessage();
+        event.partial = true;
+      }
+    }
+
+    if (FLAGS_ntfs_event_publisher_debug) {
+      TLOG << "NTFSEventPublisher event: " << event;
+    }
+
+    event_context->event_list.push_back(std::move(event));
+  }
+
+  // Put a limit on the size of the caches we are using
+  // NOTE(ww): We could also try to incrementally free up
+  // the parent FRN cache by tracking DirectoryDeletions.
+  for (auto& p : d_->reader_service_map) {
```

Comment:
You have a read lock from the stack of the function scope, but I *think* this is modifying content in `d_`. I haven't traced completely but I wanted to raise this question.

Reply:
> You have a read lock from the stack of the function scope, but I _think_ this is modifying content in `d_`.

Yep, I think your reading is correct. I didn't write this part of the event publisher, but I *think* the read lock here is intentional -- we're locking on `reader_service_map_mutex` to prevent changes to the top level `reader_service_map` when pumping events during `run`, which can only occur if `restartJournalReaderServices` fires (i.e., during `configure`), which would involve taking/waiting for the write lock.

So, I *think* the use of just a read lock here is safe, since the content being modified in `d_` can't be contended across multiple threads. But it's possibly that I'm badly misunderstanding osquery's threading/pubsub model.